### PR TITLE
Update state during learning #406

### LINKED ISSUE #406: Update state during learning
Models rewrite `state` instead of updating during learning.
Which means previous learning results are just droppe

### LINKED ISSUE #410: KeyError: 'ipeaks'
Analytics down with KeyError, while the webbooks are working,
because models don't know about `self.state` 
```
2019-02-12 12:16:37,025 [Analytics] [ERROR]  handle_analytic_task Exception: 'Traceback (most recent call last):
  File "bin/../analytics/analytic_unit_manager.py", line 94, in handle_analytic_task
    result_payload = await self.__handle_analytic_task(task)
  File "bin/../analytics/analytic_unit_manager.py", line 76, in __handle_analytic_task
    res = await worker.recieve_data(data, payload['cache'])
  File "bin/../analytics/analytic_unit_worker.py", line 41, in recieve_data
    return self._detector.recieve_data(data, cache)
  File "bin/../analytics/detectors/pattern_detector.py", line 77, in recieve_data
    res = self.detect(self.bucket.data, cache)
  File "bin/../analytics/detectors/pattern_detector.py", line 54, in detect
    detected = self.model.detect(dataframe, cache)
  File "bin/../analytics/models/model.py", line 88, in detect
    result = self.do_detect(dataframe)
  File "bin/../analytics/models/peak_model.py", line 80, in do_detect
    return self.__filter_detection(segments, data)
  File "bin/../analytics/models/peak_model.py", line 88, in __filter_detection
    if len(segments) == 0 or len(self.state['ipeaks']) == 0:
KeyError: 'ipeaks'
```

### PR DESCRIPTION
**Pull request by VargBurz**
closes #406 
closes #410

### CODE REVIEW

File: analytics/analytics/models/drop_model.py [hunk 1]
```diff
@@ -42,7 +42,7 @@ def do_fit(self, dataframe: pd.DataFrame, labeled_segments: list, deleted_segmen
         data = utils.cut_dataframe(dataframe)
         data = data['value']
         window_size = self.state['WINDOW_SIZE']
-        self.state['idrops'] = learning_info['segment_center_list']
+        self.state['idrops'] = list(set(self.state['idrops'] + learning_info['segment_center_list']))
```
Comment (amper43) [inline analytics/analytics/models/drop_model.py]: what if `self.state['idrops']` not exist? please, use `self.get('idrops', [])` to cover this case.

File: analytics/analytics/models/general_model.py [hunk 2]
```diff
@@ -40,7 +40,7 @@ def find_segment_center(self, dataframe: pd.DataFrame, start: int, end: int) ->
     def do_fit(self, dataframe: pd.DataFrame, labeled_segments: list, deleted_segments: list, learning_info: dict) -> None:
         data = utils.cut_dataframe(dataframe)
         data = data['value']
-        self.state['ipats'] = learning_info['segment_center_list']
+        self.state['ipats'] = list(set(self.state['ipats'] + learning_info['segment_center_list']))
```
Comment (amper43) [inline analytics/analytics/models/general_model.py]: same as above

File: analytics/analytics/models/drop_model.py [hunk 3]
```diff
@@ -42,10 +42,10 @@ def do_fit(self, dataframe: pd.DataFrame, labeled_segments: list, deleted_segmen
         data = utils.cut_dataframe(dataframe)
         data = data['value']
         window_size = self.state['WINDOW_SIZE']
-        self.state['idrops'] = learning_info['segment_center_list']
-        self.state['model_drop'] = utils.get_av_model(learning_info['patterns_list'])
-        convolve_list = utils.get_convolve(self.state['idrops'], self.state['model_drop'], data, window_size)
-        correlation_list = utils.get_correlation(self.state['idrops'], self.state['model_drop'], data, window_size)
+        self.state['pattern_center'] = list(set(self.state.get('pattern_center', []) + learning_info['segment_center_list']))
```
Comment (rozetko) [inline analytics/analytics/models/drop_model.py]: Let's split this expression into a few lines:
`list(set(self.state.get('pattern_center', [])))`

File: analytics/analytics/models/model.py [hunk 4]
```diff
@@ -74,10 +75,15 @@ def fit(self, dataframe: pd.DataFrame, segments: list, cache: Optional[ModelCach
                 max_length = max(segment.length, max_length)
                 if segment.labeled: labeled.append(segment)
                 if segment.deleted: deleted.append(segment)
-                    
-        self.state['WINDOW_SIZE'] = math.ceil(max_length / 2) if max_length else 0
+        if self.state.get('WINDOW_SIZE') == 240:
```
Comment (amper43) [inline analytics/analytics/models/model.py]: what mean 240 ?

File: analytics/analytics/models/model.py [hunk 5]
```diff
@@ -74,10 +75,15 @@ def fit(self, dataframe: pd.DataFrame, segments: list, cache: Optional[ModelCach
                 max_length = max(segment.length, max_length)
                 if segment.labeled: labeled.append(segment)
                 if segment.deleted: deleted.append(segment)
-                    
-        self.state['WINDOW_SIZE'] = math.ceil(max_length / 2) if max_length else 0
+        if self.state.get('WINDOW_SIZE') == 240:            
+            self.state['WINDOW_SIZE'] = math.ceil(max_length / 2) if max_length else 0
         model, model_type = self.get_model_type()
         learning_info = self.get_parameters_from_segments(dataframe, labeled, deleted, model, model_type)
+        if len(self.state.get('pattern_center', [])) > 0 and len(self.state.get('pattern_model', [])) > 0:
```
Comment (amper43) [inline analytics/analytics/models/model.py]: just `if  self.state.get('pattern_center') and self.state.get('pattern_model')`
note that `bool([])` is `False`

**Pull request merged**